torch/csrc/jit/ir_import.cpp

```diff
@@ -0,0 +1,559 @@
+  }
+
+  return subgraphs;
+}
+
+std::shared_ptr<Graph> buildGraph(const Graph_& graph_);
+
+std::vector<std::shared_ptr<Graph>> buildGraphs(const std::vector<Graph_>& graphs_);
+
+void buildBlocks(const std::vector<Graph_>& graphs_, Node* node,
+                 std::unordered_map<std::string, Value*>& value_map);
+
+void buildBlock(const Graph_& graph_, Block* block,
+                std::unordered_map<std::string, Value*>& value_map) {
+
+  for (auto input : graph_.inputs) {
+    value_map[input.name] = block->addInput();
+  }
+
+  for (auto node_ : graph_.nodes) {
+    TORCH_ASSERT(node_.op_type != "CppOp");
+    TORCH_ASSERT(node_.op_type != "PythonOp");
+
+    auto node = block->owningGraph()->create(Symbol::fromQualString(node_.op_type),
+                                             node_.outputs.size());
+
+    for (auto attr : node_.attrs) {
+      Symbol name = Symbol::attr(attr.name);
+
+      if (attr.type && onnx_AttributeProto_AttributeType_GRAPHS &&
+          attr.name == "_blocks") {
+        buildBlocks(readSubgraphs(attr.gs), node, value_map);
+        continue;
+      }
+
+      switch(attr.type) {
+        case onnx_AttributeProto_AttributeType_UNDEFINED:
+          throw std::runtime_error("UNDEFINED attribute unsupported");
+          break;
+        case onnx_AttributeProto_AttributeType_FLOAT:
+          node->f_(name, attr.f);
+          break;
+        case onnx_AttributeProto_AttributeType_INT:
+          node->i_(name, attr.i);
+          break;
+        case onnx_AttributeProto_AttributeType_STRING:
+          node->s_(name, std::move(attr.s));
+          break;
+        case onnx_AttributeProto_AttributeType_TENSOR:
+          node->t_(name, buildTensor(attr.t));
+          break;
+        case onnx_AttributeProto_AttributeType_GRAPH:
+          node->g_(name, buildGraph(readSubgraph(attr.g)));
+          break;
+        case onnx_AttributeProto_AttributeType_FLOATS:
+          node->fs_(name, std::move(attr.fs));
+          break;
+        case onnx_AttributeProto_AttributeType_INTS:
+          node->is_(name, std::move(attr.is));
+          break;
+        case onnx_AttributeProto_AttributeType_STRINGS:
+          node->ss_(name, std::move(attr.ss));
+          break;
+        case onnx_AttributeProto_AttributeType_TENSORS:
+          node->ts_(name, buildTensors(attr.ts));
+          break;
+        case onnx_AttributeProto_AttributeType_GRAPHS:
+          node->gs_(name, buildGraphs(readSubgraphs(attr.gs)));
+          break;
+      }
+    }
+
+    for (auto input : node_.inputs) {
+      auto v = value_map[input];
+      node->addInput(v);
+    }
+
+    for (size_t i=0; i<node_.outputs.size(); i++) {
+      value_map[node_.outputs[i]] = node->outputs()[i];
+    }
+
+    block->appendNode(node);
+  }
+
+  for (auto output : graph_.outputs) {
+    if (value_map.find(output.name) != value_map.end()) {
+      Value* v = value_map.at(output.name);
+      block->registerOutput(v);
+    }
```

Comment:
Why do we silently ignore outputs that are not in `value_map`? That looks like an error to me

Reply:
Shouldn't all values at this point be in `value_map`? The inputs to the graph and the outputs from every node have been stored in `value_map` earlier, so they have to be there.
Maybe an `assert` would be more appropriate.

Reply:
They should. My point is that if they weren't then you'd silently ignore them. The code might work now, but this makes it really easy to add some bugs in the future and don't even notice them (hey, it works!).